### PR TITLE
Major Release v1.0

### PR DESCRIPTION
**Pull request by Zilinghan**
A complete refactor of APPFL with the following main changes:
(1) Improve the modularity and extensibility of the framework by using server and client agents
(2) Separate aggregation algorithms, scheduling algorithms, and local training algorithms for better modularity
(3) Incorporate compressor for all communication manners
(4) Support asynchronous communications for gRPC
(5) Support authentication for gRPC
(6) Rewrite MPI server into a request-response server for easy user customization
(7) Add full documentation including installation, tutorial, user guide, and developer guide

### CODE REVIEW

File: docs/developer/algorithms.rst
Comment (kibaekkim) [inline docs/developer/algorithms.rst:1]: Describe (or enumerate) a list of aggregators available in the package.

File: setup.py [hunk 1]
```diff
@@ -47,23 +52,30 @@
     ],
     package_dir={"": "src"},
     packages=setuptools.find_packages(where="src"),
-    python_requires=">=3.6",
+    include_package_data=True, 
+    package_data={
+        "": ["*.sh", "*.crt", "*.key"], 
+    },
+    python_requires=">=3.8",
     install_requires=[
```
Comment (kibaekkim) [inline setup.py:60]: Dependent packages
- I think most users would not need globus. The packages related to globus can be optional.
- Do we need `torchvision` and `matplotlib` as required packages?
- `tensorboard` is removed. Do we not support it anymore?

File: .github/workflows/build.yml
Comment (kibaekkim) [inline .github/workflows/build.yml:1]: The test part has been commented out. We need some test scripts.

File: setup.py [hunk 2]
```diff
@@ -47,23 +52,30 @@
     ],
     package_dir={"": "src"},
     packages=setuptools.find_packages(where="src"),
-    python_requires=">=3.6",
+    include_package_data=True, 
+    package_data={
+        "": ["*.sh", "*.crt", "*.key"], 
+    },
+    python_requires=">=3.8",
```
Comment (kibaekkim) [inline setup.py:59]: The requirement is changaed from 3.6 to 3.8. Any reasons for higher versions?

File: docs/tutorials/index.rst
Comment (kibaekkim) [inline docs/tutorials/index.rst:1]: In `tutorials` the documentations for Google Cloud and Globus have been removed. Do we have the documentations somewhere else in the doc? We probably don't need/want to maintain the Google Cloud ones. But Globus documentation would be useful.

File: docs/developer/algorithms.rst
Comment (kibaekkim) [reply docs/developer/algorithms.rst:1]: It's under `users`, which makes sense.

File: docs/index.rst
Comment (kibaekkim) [inline docs/index.rst:1]: Do we have a documentation for the configuration? We need descriptions for the supported configuration parameters and user-defined ones.

Comment (Zilinghan) [reply docs/index.rst:1]: A notebook: https://github.com/APPFL/APPFL/blob/zilinghan/grpc-update/docs/notebooks/configuration.ipynb
Detailed descriptions of configurations for server and clients: https://github.com/APPFL/APPFL/blob/zilinghan/grpc-update/docs/users/server_agent.rst#client-configurations and https://github.com/APPFL/APPFL/blob/zilinghan/grpc-update/docs/users/client_agent.rst#configurations

File: docs/tutorials/index.rst
Comment (Zilinghan) [reply docs/tutorials/index.rst:1]: We have a new documentation for AWS: https://github.com/APPFL/APPFL/blob/zilinghan/grpc-update/docs/tutorials/aws.rst

File: setup.py (hunk 2 above)
Comment (Zilinghan) [reply setup.py:59]: Globus Compute requires a higher version of python

File: src/appfl/agent/server.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+            model_configs.model_name,
+            **(model_configs.model_kwargs if hasattr(model_configs, "model_kwargs") else {})
+        )
+        # load the model source file and delete model path
+        if hasattr(self.server_agent_config.client_configs, "model_configs"):
+            with open(model_configs.model_path, 'r') as f:
+                self.server_agent_config.client_configs.model_configs.model_source = f.read()
+            del self.server_agent_config.client_configs.model_configs.model_path
+
+    def _load_loss(self) -> None:
```
Comment (kibaekkim) [inline src/appfl/agent/server.py:238]: To my understanding, this needs users to create a separate file that defines a loss function class. 
- Do users need to follow a certain form for the class? In example, the loss function class defines `__init__` and `forward`. Is the `forward` method required to implement?
- Is there a way for users to simply indicate pytorch function?

Comment (Zilinghan) [reply src/appfl/agent/server.py:238]: As shown [here](https://github.com/APPFL/APPFL/blob/zilinghan/grpc-update/docs/users/user_loss.rst), user need to define loss as a `torch.nn.Module` with a defined forward function to compute the loss. Compared with a simple pytorch function, this gives user some flexibility to have some additional arguments when intializing the loss function, e.g., `CrossEntroyLoss(reduction='mean')`.

File: .github/workflows/build.yml
Comment (Zilinghan) [reply .github/workflows/build.yml:1]: Will do soon

File: setup.py (hunk 1 above)
Comment (Zilinghan) [reply setup.py:60]: 1. It can be optional, but as all globus sdks are very light-weight, do you think it is really necessary to mark them optional.
2. Great catch! `torchvision` is not required and I will move it to `examples`. `matplotlib` is required as it is used in some misc functions within `appfl.misc`
3. I will add tensorboard back, some codes still support tensorboard

File: src/appfl/algorithm/scheduler/base_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+from omegaconf import DictConfig
+from concurrent.futures import Future
+from typing import Union, Dict, Any, Tuple, OrderedDict
+
+class BaseScheduler:
+    def __init__(
+        self, 
+        scheduler_configs: DictConfig, 
+        aggregator: Any,
+        logger: Any
```
Comment (kibaekkim) [inline src/appfl/algorithm/scheduler/base_scheduler.py:11]: Can we specify types of these attributes?

File: src/appfl/algorithm/scheduler/sync_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        """
+        Schedule a synchronous global aggregation for the local model from a client.
+        The method will return a future object for the aggregated model, which will
+        be set after all clients have submitted their local models for the global aggregation.
+        :param client_id: the id of the client
+        :param local_model: the local model from a client
+        :param kwargs: additional keyword arguments for the scheduler
+        :return: the future object for the aggregated model
+        """
+        future = Future()
```
Comment (kibaekkim) [inline src/appfl/algorithm/scheduler/sync_scheduler.py:29]: Where do we set the number of threads? Does this assume that the server always runs on CPU?

File: src/appfl/agent/server.py (hunk 3 above)
Comment (kibaekkim) [reply src/appfl/agent/server.py:238]: Can you docuement this, if not already done?

Comment (Zilinghan) [reply src/appfl/agent/server.py:238]: Here the the document for this https://github.com/APPFL/APPFL/blob/zilinghan/grpc-update/docs/users/user_loss.rst

File: src/appfl/algorithm/scheduler/base_scheduler.py (hunk 4 above)
Comment (Zilinghan) [reply src/appfl/algorithm/scheduler/base_scheduler.py]: Yes, I changed to `BaseAggregator`

File: src/appfl/algorithm/scheduler/sync_scheduler.py (hunk 5 above)
Comment (Zilinghan) [reply src/appfl/algorithm/scheduler/sync_scheduler.py:30]: This does not assume the server has to run on CPU. The number of threads supported depends on the communicator used, for example, for grpc, the threads are specified here: https://github.com/APPFL/APPFL/blob/63810d2350dc51c1f3f64a031f3f1a915e7c13f8/src/appfl/comm/grpc/serve.py#L26

File: .github/workflows/build.yml
Comment (Zilinghan) [reply .github/workflows/build.yml:1]: @kibaekkim The tests are added!

File: src/appfl/algorithm/scheduler/sync_scheduler.py (hunk 5 above)
Comment (Zilinghan) [reply src/appfl/algorithm/scheduler/sync_scheduler.py]: Hi @kibaekkim, here are some conclusions I have regarding the usage of `concurrent.futures.Future`:

**TL;DR**: The usage of the Future object in all the schedulers does not restrict the server device to CPU-only, and there is no need to set a `max_workers` for that.

**Detailed Step-by-step Explanations**:

(1) What is the relationship between `concurrent.futures.Future` and `concurrent.futures.ThreadPoolExecutor`?

Answer: `ThreadPoolExecutor(max_workers=10)` provides an executor that you can submit as many concurrent functions to run as you want in an asynchronous way (note: at most 10 running concurrently, and if you submit more at the same time, some will be queued), and the running of these functions do not block the main program. As a consequence, the result of `executor.submit` is of type `Future`, and will be set to the actual results when the asynchronous function finishes running.

(2) If we directly create a `Future` object and set its value, does any concurrency involved?

Answer: No, for example, for the following code, everything runs serially.
```python
future = Future()
print(future.done()) # False - as no result is set for it
future.set_result(42)
print(future.done()) # True
print(future.result()) # 42
```

(3) How does the scheduler utilize the `concurrent.futures.Future` object?

Answer: Same as the above example, the scheduler module creates a `Future` object to indicate that the results (i.e., the global model) are not yet available - there are no concurrency/multi-thread involved in this process.

(4) Okay, so no concurrency involved, why we have `futures.ThreadPoolExecutor(max_workers=max_workers)` within the gRPC server?

Answer: gRPC uses this `ThreadPoolExecutor` to launch multiple threads to be able to handle requests from multiple clients at the same time. In this case, each thread gets a `Future` object after calling the sync scheduling function.

(5) Why the MPI server does not need multiple threads?

Answer: In our MPI communicator, instead of handling request for multiple clients at the same time, it receives client’s local model one by one from its receiving queue, passes to the scheduler, and gets a `Future` object. All those `Future` objects will be set after receiving the last model and performing the global aggregation. The set `Future` objects (i.e., new global model) will be send back to clients.

(6) So gRPC server uses multi thread while MPI server does not, does this affect the efficiency?

Answer: No for two reasons.

- All the scheduling logics have critical section that are protected by locks (e.g., [sync](https://github.com/APPFL/APPFL/blob/b3932ea32874c0d51ac44c12b4d3d3bec309079d/src/appfl/algorithm/scheduler/sync_scheduler.py#L33) and should be executed serially, so even with multiple threads, the scheduler runs serially.
- The scheduling can be described as IO-bound, i.e., its efficiency depends on when the last model will be sent to the server, so the concurrency will not have large increase to the overall efficiency.

**Review by kibaekkim — APPROVED** [review]
Excellent work @Zilinghan !

I am approving the PR with a couple TODOs for future:
- There are warnings from torch, as you may know. Please plan to address this.
```
FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
```
- Receiving client messages in series in MPI may degrade the performance, particularly when the clients arrive about the same time. Please consider to add a collective communication such as gather.
